src/core/qgsgeometryvalidator.cpp

```diff
@@ -315,18 +317,20 @@ void QgsGeometryValidator::run()
         }
 
         case QgsWkbTypes::Polygon:
-          validatePolygon( 0, qgsgeometry_cast< const QgsPolygon * >( mGeometry.constGet() ) );
+        case QgsWkbTypes::CurvePolygon:
+          validatePolygon( 0, qgsgeometry_cast< const QgsCurvePolygon * >( mGeometry.constGet() ) );
           break;
 
         case QgsWkbTypes::MultiPolygon:
+        case QgsWkbTypes::MultiSurface:
         {
           const QgsGeometryCollection *collection = qgsgeometry_cast< const QgsGeometryCollection * >( mGeometry.constGet() );
           for ( int i = 0; !mStop && i < collection->numGeometries(); i++ )
-            validatePolygon( i, qgsgeometry_cast< const QgsPolygon * >( collection->geometryN( i ) ) );
+            validatePolygon( i, qgsgeometry_cast< const QgsCurvePolygon * >( collection->geometryN( i ) ) );
```

Comment:
side remark, a bit beyond the scope of this PR: if instead of casting to a QgsGeometryCollection but to a QgsMultiSurface, and if QgsMultiSurface would override geometryN() to return a QgsCurvePolygon* instead of a QgsAbstractGeometry*, the casts at that line at at line 333 could be removed

Reply:
Right. With @Koyaani we're working on some improvements (and fixes) like this. Thanks